sam.c

```diff
@@ -1924,6 +1937,614 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
+                char *rp = realloc(l->data, l->alloc * 2);
+                if (!rp) {
+                    fd->errcode = ENOMEM;
+                    goto tidyup;
+                }
+                l->alloc *= 2;
+                l->data = rp;
+                assert(l->alloc >= l->data_size);
+                assert(l->alloc >= line_frag);
+                assert(l->alloc >= l->alloc - line_frag);
+                goto longer_line;
+            }
+            cp++;
+
+            // line holds the remainder of our line.
+            ks_resize(&line, cp_end - cp);
+            memcpy(line.s, cp, cp_end - cp);
+            line_frag = cp_end - cp;
+            l->data_size = l->alloc - line_frag;
+        } else {
+            // out of buffer
+            line_frag = 0;
+        }
+
+        l->serial = fd->serial++;
+        //fprintf(stderr, "Dispatching %p, %d bytes, serial %d\n", l, l->data_size, l->serial);
+        hts_tpool_dispatch(fd->p, fd->q, sam_parse_worker, l);
+    }
+
+    hts_tpool_dispatch(fd->p, fd->q, sam_parse_eof, NULL);
+
+    // At EOF, but there is potential for the caller to then do another seek.
+    for (;;) {
+        pthread_mutex_lock(&fd->command_m);
+        if (fd->command == SAM_NONE)
+            pthread_cond_wait(&fd->command_c, &fd->command_m);
+        switch (fd->command) {
+        case SAM_CLOSE:
+            pthread_cond_signal(&fd->command_c);
+            pthread_mutex_unlock(&fd->command_m);
+            hts_tpool_process_destroy(fd->q);
+            fd->q = NULL;
+            goto tidyup;
+
+        default:
+            pthread_mutex_unlock(&fd->command_m);
+            break;
+        }
+    }
+
+ tidyup:
+    if (l) {
+        l->next = fd->lines;
+        fd->lines = l;
+    }
+    free(line.s);
+
+    return NULL;
+}
+
+// Runs in the main thread.
+// Takes encoded blocks of SAM off the thread results queue and writes them
+// to our output stream.
+static void *sam_dispatcher_write(void *vp) {
+    htsFile *fp = vp;
+    SAM_state *fd = fp->state;
+    hts_tpool_result *r;
+
+    // Iterates until result queue is shutdown, where it returns NULL.
+    while ((r = hts_tpool_next_result_wait(fd->q))) {
+        sp_lines *gl = (sp_lines *)hts_tpool_result_data(r);
+        if (!gl)
+            continue;
+
+        if (fp->idx) {
+            sp_bams *gb = gl->bams;
+            int i = 0, count = 0;
+            while (i < gl->data_size) {
+                int j = i;
+                while (i < gl->data_size && gl->data[i] != '\n')
+                    i++;
+                if (i < gl->data_size)
+                    i++;
+
+                if (fp->format.compression == bgzf) {
+                    if (bgzf_write(fp->fp.bgzf, &gl->data[j], i-j) != i-j) {
+                        pthread_mutex_lock(&fd->command_m);
+                        fd->errcode |= BGZF_ERR_IO;
+                        pthread_mutex_unlock(&fd->command_m);
+                        fd->errcode |= BGZF_ERR_IO;
```

Comment:
Odd duplicated line here.  Presumably the one that updates `fd->errcode` with the mutex locked is the right one.  There are other example of this elsewhere.  Also, `fd->errcode` is not consistently locked everywhere.  Maybe a static inline to access it safely would be a good idea?

Finally, it sometimes gets `errno` values (e.g. `EIO`), and others (like here) `BGZF_ERR_IO`.  Which should it be?

Reply:
I'll use EIO, ENOMEM, etc and avoid the BGZF_ERR codes as they're a bit unhelpful.

However arguably it's redundant as there's no way to return errors other than pass and fail from the high level functions anyway.  Still, it's there now so I may as well keep it and we can figure out how best to deal with global errors later.  (Maybe just set the global errno at the point of top-level failing would be best.)

I did try to figure out the return codes for some of our functions, but they're mostly undocumented with <0 being error, despite actually having different values for different things.